compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1810,6 +1810,10 @@ class Typer extends Namer
       // check value class constraints
       checkDerivedValueClass(cls, body1)
 
+      val effectiveOwner = cls.owner.skipWeakOwner
+      if cls.is(CaseClass) && effectiveOwner.is(Trait) && !effectiveOwner.derivesFrom(defn.ObjectClass)
```

Comment:
I think the previous version of this check was closer to being correct: defining a class or an object in a value class is disallowed, so we should similarly disallow them in universal traits. There was one build failure because the universal trait SeqOps in the standard library defines some private inner classes, which I think is fine as long as we infer that those are `private[this]` classes: 
```suggestion
      if !cls.isRefinementClass && cls.isAllOf(PrivateLocal) && effectiveOwner.is(Trait) &&
         !effectiveOwner.derivesFrom(defn.ObjectClass)
      then
```